Distributed tracing 2.9 (#62410)
openshift/openshift-docs: distr_tracing/distr_tracing_otel/distr-tracing-otel-forwarding.adoc

:_content-type: PROCEDURE
[id="distr-tracing-otel-forwarding"]
= Forwarding traces to TempoStack by using OpenTelemetry Collector
include::_attributes/common-attributes.adoc[]
:context: distr-tracing-otel-forwarding

> To configure log forwarding to the TempoStack, you deploy and configure an OpenTelemetry collector.

Reviewer: This should be either.

To configure forwarding Telemetry data..
or
To configure forwarding traces...

> To configure log forwarding to the TempoStack, you deploy and configure an OpenTelemetry collector.

Author: @IshwarKanse, like this? 
```suggestion
To configure forwarding telemetry data to the TempoStack, you deploy and configure an OpenTelemetry collector.
```

.Prerequisites

* Tempo Operator is installed.
* TempoStack is deployed on the cluster.

> * Red Hat OpenShift distributed tracing data collection operator is installed.

Reviewer 2: Should this be pointing to variables?

> .Prerequisites
> 
> * Tempo Operator is installed.
> * TempoStack is deployed on the cluster.
> * Red Hat OpenShift distributed tracing data collection operator is installed.

Author: @rubenvp8510, I suggest removing the prerequisites section completely because these prerequisites are self-explanatory, which means we don't have to mention them: 
```suggestion
.Prerequisites

* Tempo Operator is installed.
* TempoStack is deployed on the cluster.
* Red Hat OpenShift distributed tracing data collection operator is installed.
```

.Procedure

> For the following steps, we will deploy an OpenTelemetry collector using the recomended processors/receivers/explorters, this example will deploy the collector in deployment mode, for other modes see the OpenTelemetry collector documentation.

Reviewer 2: ```suggestion
For the following steps, we will deploy an OpenTelemetry collector using the recomended processors/receivers/exporters, this example will deploy the collector in deployment mode, for other modes see the OpenTelemetry collector documentation.
```

. Create a service account for the OpenTelemetry collector, here is an example:
+
----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: otel-collector-deployment
----

. Create a cluster role, this cluster role will  be associated with the service account created in the first step:
+
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: otel-collector
rules:
# Permissions for pods and namespaces resources are needed for the k8sattributesprocessor
# Permissions for infrastructures and infrastructures/status are needed for resourcedetectionprocessor
- apiGroups: ["", "config.openshift.io"]
  resources: ["pods", "namespaces", "infrastructures", "infrastructures/status"]
  verbs: ["get", "watch", "list"]
----

> Some permissions are neccesary to make the `k8sattributesprocessor` and `resourcedetectionprocessor` works, and those are recomended.

Reviewer 2: Maybe this or the comments in the CR could be removed

roleRef:
  kind: ClusterRole
  name: otel-collector
  apiGroup: rbac.authorization.k8s.io
----

. Create the YAML file that will define the `OpenTelemetryCollector` custom resource (CR):
+
----
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otel
spec:
  mode: deployment
  serviceAccount: otel-collector-deployment
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
          thrift_binary:
          thrift_compact:
          thrift_http:
      opencensus:
      otlp:
        protocols:
          grpc:
          http:
      zipkin:
    processors:
      batch:
      k8sattributes:
      memory_limiter:
        check_interval: 1s
        limit_percentage: 50
        spike_limit_percentage: 30
      resourcedetection:
        detectors: [openshift]
    exporters:
      otlp:
        endpoint: "tempo-simplest-distributor:4317" # Endpoint for the tempo distributor.
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [jaeger, opencensus, otlp, zipkin]
          processors: [memory_limiter, k8sattributes, resourcedetection, batch]
          exporters: [otlp]
----

> The collector is configured with a receiver for Jaeger traces, OpenCensus traces over the OpenCensus protocol, Zipkin traces over the Zipkin protocol and OTLP traces over the GRPC protocol. The collector exporter is configured to export OTLP, and is pointing to the tempo distributor endpoint, in this case "tempo-simplest-distributor:4317" which should be already created.

Reviewer 2: ```suggestion
The collector is configured to receive traces in Jaeger, OpenCensus, Zipkin and format over different protocols. The collector exporter is configured to export traces using the OTLP protocol, and is pointing to the tempo distributor endpoint, in this case "tempo-simplest-distributor:4317" which should be already created.
```

> You can find more examples of how to deploy OpenTelemetry with tempo in sidecar mode and multitenant here: https://github.com/os-observability/redhat-rhosdt-samples

Reviewer 2: ```suggestion
You can find more examples of how to deploy OpenTelemetry as a sidecar or using multitenancy to forward traces to Tempo here: https://github.com/os-observability/redhat-rhosdt-samples
```
